src/schema/v2/filterArtworksConnection.ts

```diff
@@ -575,6 +572,9 @@ const filterArtworksConnectionTypeFactory = (
         )
       }
 
+      // temp
+      hits.forEach((artwork) => enrichArtworkWithCollectorSignals(artwork, ctx))
```

Comment:
This feels problematic - a call to `enrich` for every artwork (often up to 30 for a full artwork grid) - each of which might fan out to _multiple_ fetches - this would block the initial request for a page featuring a grid (artist page, collect/collection page).

Depending on the approach taken - since eventually collector signals _would_ be asked for on every artwork grid/rail - it might be nice to move everything about collector signals to the client - maybe a bespoke endpoint that returns all the collector signals for a list of artwork id's which can be queried + rendered on the client - or perhaps queried all at once.

As-is this probably needs an `await` or something here as the `enrich` is async.

Reply:
Oh you are right. that might explain some more empty results. I just wanted to see it working as inefficiently as possible. I'll change this to a `await Promise.all(hits.map(...)`.

Regarding doing it on the client vs here, that's what I am wondering too. We have a hook there.

Reply:
Looking at Datadog ([example](https://app.datadoghq.com/apm/trace/4639396345898945969?colorBy=service&env=production&graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=4234249432811012679&spanViewType=metadata&timeHint=1721380110116.001&view=spans)), it seems the API call to Gravity filter endpoint is one of the main bottle neck:

<details><summary>Screenshot</summary>
<p>

![Screenshot 2024-07-19 at 9 34 04 AM](https://github.com/user-attachments/assets/e223fe34-770f-45b2-b005-4aa2d1608af0)

</p>
</details> 

Additional requests that got fanned out afterwards are surprisingly parallelized and not so bad:

<details><summary>Screenshot</summary>
<p>

![Screenshot 2024-07-19 at 9 34 38 AM](https://github.com/user-attachments/assets/bb43e686-1d2a-4c3e-bd9d-fb1f75ed96d5)

</p>
</details> 

I totally agree that moving them to client side can help with unblocking rendering (and we'll need to raise it to design for the right placeholder patterns for this case), but I also feel the exact data fetched here (like bid count etc.) have good opportunities to be optimized (e.g. cached or even persisted).

To help us learn and determine the right next steps, maybe we can put this behind a feature flag (following the [blur hash pattern](https://github.com/artsy/metaphysics/blob/900d031a5df9e7ec3ef89412039c30d62c8f8c2f/src/schema/v2/image/index.ts#L79)) and we can observe the performance with real/staging data?

In the mean time, I'm going to give design a heads-up and get quick reaction re rendering the signals on the client-side.

Reply:
Definitely agree - anything iterable/behind a feature flag is a good idea and 👍 from me.

Also yup good call out - client-side rendering does have that subtle design consideration of needing to likely support the loading state.

That chart is cool! Makes sense the ES call that front-loads any fanning out of enriching artwork calls is the slowest (and the parallelization of multiple smaller cacheable requests might be _more_ performant than one bigger batched endpoint that returns all signals for a list of artworks, for instance).

This further enriching behind a feature flag definitely makes this totally mergeable.